### PR TITLE
Export with XAI branch

### PR DESCRIPTION
**Pull request by negvet**
### Summary

Support XAI export

Manually validated for 
- multi_class_cls/mobilenet_v3_large_light
- multi_class_cls/efficientnet_b0_light
- multi_class_cls/efficientnet_v2_light

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/model/entity/classification.py [hunk 1]
```diff
@@ -80,6 +80,33 @@ def reset_explain_hook(self) -> None:
         """Clear all history of explain records."""
         self.explain_hook.reset()
 
+    def get_custom_forward(self):
+        from otx.algo.hooks.recording_forward_hook import ReciproCAMHook
```
Comment (sovrasov) [inline src/otx/core/model/entity/classification.py:88]: get_forward_with_explain()? Or we can add forward_with_explain() as an abstract method and suppose we always have it. If so, we don't need to pass it to the exporter as a parameter.
Does that impl work for all cls models?

File: src/otx/core/exporter/base.py [hunk 2]
```diff
@@ -211,3 +220,15 @@ def _embed_openvino_ir_metadata(ov_model: openvino.Model, metadata: dict[tuple[s
...
         return ov_model
+
+    def _patch_model(self, model):
+        if self.custom_forward:
+            funcType = types.MethodType
+            model.forward = funcType(self.custom_forward, model)
+
+    def _update_output_names(self):
+        if "output_names" not in self.onnx_export_configuration:
+            self.onnx_export_configuration["output_names"] = ["logits", "saliency_map"]
```
Comment (sovrasov) [inline src/otx/core/exporter/base.py]: Not all the model are exported to IRs via onnx, so that statement may have no effect.
As a universal way (for OVC and torch.onnx), one can return a dict with the required keys from the custom forward.

File: src/otx/core/exporter/base.py [hunk 3]
```diff
@@ -64,6 +67,7 @@ def export(
         base_model_name: str = "exported_model",
         export_format: OTXExportFormatType = OTXExportFormatType.OPENVINO,
         precision: OTXPrecisionType = OTXPrecisionType.FP32,
+        dump_auxiliaries: bool = False,
```
Comment (sovrasov) [inline src/otx/core/exporter/base.py]: We may remove this, since custom_forward can be optional, and can be applied only when it's actually passed.

File: src/otx/core/model/entity/base.py [hunk 4]
```diff
@@ -248,9 +249,14 @@ def _export_parameters(self) -> dict[str, Any]:
             ("model_info", "label_ids"): all_label_ids.strip(),
             ("model_info", "optimization_config"): json.dumps(optimization_config),
         }
+        parameters["custom_forward"] = self.get_custom_forward()
 
         return parameters
 
+    def get_custom_forward(self):
```
Comment (sovrasov) [inline src/otx/core/model/entity/base.py]: I'd keep it as an private property. Can we consider renaming this to `_explain_forward` for clarity?

File: src/otx/core/model/entity/base.py [hunk 5]
```diff
@@ -213,7 +214,7 @@ def export(
         Returns:
             Path: path to the exported model.
         """
-        return self._exporter.export(self.model, output_dir, base_name, export_format, precision)
+        return self._exporter.export(self.model, output_dir, base_name, export_format, precision, dump_auxiliaries)
```
Comment (sovrasov) [inline src/otx/core/model/entity/base.py]: When dump_auxiliaries is true, we need a check here, if custom forward actually exists or not 
If not -> not implemented error

File: src/otx/core/exporter/base.py (hunk 3 above)
Comment (negvet) [reply src/otx/core/exporter/base.py]: Exporter constructor will always get implementation of `custom_forward`, however it can be None.
It is not that handy for the user to control dumping feature vector and saliency maps (auxiliaries) by the availability of `custom_forward`. For models that support dumping auxiliaries, there always gonna be `custom_forward` implemented, but still the user should have the options to disable dumping auxiliaries (it should be by default actually).

Comment (sovrasov) [reply src/otx/core/exporter/base.py]: I see, it looks like controlling forward is a bit painful in the exporter, so maybe we move it to the OTXModel?
This'd allow overloading exporters with model patching, and also this reduces the depth of modifications, when we need to change something related to model customization.

Comment (negvet) [reply src/otx/core/exporter/base.py]: Ok, let me see what I can do

File: src/otx/core/exporter/base.py (hunk 2 above)
Comment (negvet) [reply src/otx/core/exporter/base.py]: Fixed, please take a look

File: src/otx/core/exporter/base.py (hunk 3 above)
Comment (negvet) [reply src/otx/core/exporter/base.py]: Moved it all to OTXModel

File: src/otx/core/model/entity/base.py (hunk 4 above)
Comment (negvet) [reply src/otx/core/model/entity/base.py]: Fixed and renamed

File: src/otx/core/model/entity/base.py (hunk 5 above)
Comment (negvet) [reply src/otx/core/model/entity/base.py]: Done

File: src/otx/core/model/entity/base.py [hunk 6]
```diff
@@ -263,6 +280,33 @@ def _reset_prediction_layer(self, num_classes: int) -> None:
...
+            msg = "original_model_forward is None."
+            raise RuntimeError(msg)
+
+        func_type = types.MethodType
+        self.model.forward = func_type(self.original_model_forward, self.model)
+        self.original_model_forward = None
+
+    @staticmethod
+    def _update_onnx_output_names(onnx_export_configuration: dict) -> None:
+        if "output_names" not in onnx_export_configuration:
```
Comment (sovrasov) [inline src/otx/core/model/entity/base.py]: This implementation is suitable only for classification models, so it should be moved to the corresponding class.

File: src/otx/core/model/entity/base.py [hunk 7]
```diff
@@ -210,10 +213,20 @@ def export(
...
+            dump_auxiliaries (bool): Flag to return "feature_vector" and "saliency_map". Defaults to False.
+
         Returns:
             Path: path to the exported model.
         """
-        return self._exporter.export(self.model, output_dir, base_name, export_format, precision)
+        exporter = self._exporter
+        if dump_auxiliaries:
+            self._reset_model_forward()
+            if hasattr(exporter, "onnx_export_configuration"):
```
Comment (sovrasov) [inline src/otx/core/model/entity/base.py:224]: Now this check looks as a workaround, which arises because `_exporter` is configurable only from the object state.
In the future, we can introduce explain mode, which affects the object state and allows switching to explain forward & export parameters.
For instance, `OTXModel.set_explain_mode(enabled=True/False)`.
That'd allow to use that mode with `litModule.predict` as well to dump explain results into a dataset.

File: src/otx/core/model/entity/base.py (hunk 6 above)
Comment (negvet) [reply src/otx/core/model/entity/base.py]: Right, fixed

File: src/otx/core/model/entity/base.py [hunk 8]
```diff
@@ -251,6 +266,10 @@ def _export_parameters(self) -> dict[str, Any]:
 
         return parameters
 
+    def _get_forward_with_auxiliaries(self) -> Callable | None:
```
Comment (vinnamkim) [inline src/otx/core/model/entity/base.py:269]: Why don't we just add `forward_explain()`  in this PR? It seems getting dirty without this I think.

Comment (sovrasov) [reply src/otx/core/model/entity/base.py:269]: If we go this way, we also have to provide a native mechanism for setting explain mode for the model. Now, it's done with a hook, but this mechanism has nothing to do with export and `forward_explain()`. So I believe it should be like explain_mode set -> forward is replaced with `forward_explain()`. At the same time, this should not be conflicting with the existing loss/predict modes.

Comment (sovrasov) [reply src/otx/core/model/entity/base.py:269]: Now it looks like custom forward immediately breaks mm's predict/loss method, while hooks allows to leave it as is.
Having this in mind, Evgeny's solution looks appropriate (replacing forward with xai one only for export), unless we change forward -> predict/loss -> _customize_outputs pipeline to support XAI natively

**Comment by sovrasov** [comment]
@negvet I believe https://github.com/openvinotoolkit/training_extensions/pull/2905/commits/f846fe551ca42900bdf4322537fe9c7d67af398f is useful for the future work

**Comment by sovrasov** [comment]
@negvet also please have a look at https://github.com/openvinotoolkit/training_extensions/pull/2913

File: src/otx/core/model/entity/base.py (hunk 8 above)
Comment (vinnamkim) [reply src/otx/core/model/entity/base.py:269]: No, I just wanna add `forward_explain()`, not overriding `forward()`. Then, use this function for 1) `OTXModel.export(with_xai=True)` and 2) `Engine.predict(with_xai=True)`.

**Comment by sovrasov** [comment]
@negvet can we close this PR?

**Comment by sovrasov** [comment]
Implemented in https://github.com/openvinotoolkit/training_extensions/pull/2918
